Framework/Algorithms/inc/MantidAlgorithms/ReflectometryReductionOneAuto2.h

```diff
@@ -82,6 +78,10 @@ class DLLExport ReflectometryReductionOneAuto2
   std::tuple<API::MatrixWorkspace_sptr, std::string, std::string>
   getPolarizationEfficiencies();
   void applyPolarizationCorrection(std::string const &outputIvsLam);
+  API::MatrixWorkspace_sptr getFloodWorkspace();
+  void applyFloodCorrection(API::MatrixWorkspace_sptr const &flood,
```

Comment:
Preferred is to avoid modifications of the MatrixWorkspace pointed to by flood. But now, you prevent modifying the pointer itself but not the MatrixWorkspace it points to.
I still would use MatrixWorkspace_const_sptr.

Reply:
oh, maybe you need to modify the MatrixWorkspace pointed to by flood here!

Reply:
In principle it should be `MatrixWorkspace_const_sptr` but I don't think `setProperty(...)` will accept it. Could you try it on your end? It didn't work for me.

Reply:
This compiles for me:
```
void ReflectometryReductionOneAuto2::applyFloodCorrection(
    MatrixWorkspace_const_sptr &flood, std::string const &prop) {
  MatrixWorkspace_sptr ws = getProperty(prop);
  auto alg = createChildAlgorithm("ApplyFloodWorkspace");
  alg->initialize();
  alg->setProperty("InputWorkspace", ws);
  alg->setProperty("FloodWorkspace", flood);
  alg->execute();
  MatrixWorkspace_const_sptr out = alg->getProperty("OutputWOrkspace");
  setProperty(prop, out);
}

void ReflectometryReductionOneAuto2::applyFloodCorrections() {
  MatrixWorkspace_const_sptr flood = getFloodWorkspace();
  if (flood) {
    applyFloodCorrection(flood, "InputWorkspace");
    if (!isDefault("FirstTransmissionRun")) {
      applyFloodCorrection(flood, "FirstTransmissionRun");
    }
    if (!isDefault("SecondTransmissionRun")) {
      applyFloodCorrection(flood, "SecondTransmissionRun");
    }
  }
}
```

Be aware, I added more consts

Reply:
I see an unhandled exception when running the tests:
```
[Error] MatrixWorkspace::getDetector(): No detectors for this workspace index. search object
```

Reply:
Maybe there is a need of something like
```
  template <typename T>
  IPropertyManager *setProperty(const std::string &name,
                                std::unique_ptr<const T> value) {
    setTypedProperty(
        name, std::move(value),
        std::is_convertible<std::unique_ptr<T>, boost::shared_ptr<DataItem>>());
    this->afterPropertySet(name);
    return this;
  }
```
in IPropertyManager.h ? Can this work?

Reply:
Or maybe use
```
setProperty(const std::string &name, const T &value)
```
so
```
setProperty(prop, &out);
```
?

Reply:
As far as I remember const workspace properties have never worked. If it's still the case we shouldn't be making any changes to Mantid's core classes as it's beyond the scope of this issue.

Reply:
ok